Development

NAbelSmith opened the pull request:
All changes included for assessment, pending reviews.

yipped reviewed (commented):
Nathan Smith's code has met the requirements of the assignment with extra features such as continuous rolling. The program runs fine and as expected on my system. The program follows a consistent naming scheme and is thoroughly commented, describing each step of every function, making it very easy to understand. All variables are checked including the random number generation.

The only very minor improvement I can think of is to reduce code reuse by creating another function for rolling die and displaying its value in Game.cs starting at line 56. For example a function called RollDie() (or something similar), where a given die object is passed through along with its name or number to use its roll method, generate text saying what the value of the roll was and return roll value.

WonderboyMattT commented on CMP1903_A1_2324/Die.cs:
The die class meets the requirements of the brief.
"_value" hold the current value of the die and is accessed through "value".
The "Roll" method generates a random in and assign it.

WonderboyMattT commented on CMP1903_A1_2324/Game.cs:
The implementation here is comprehensive and meets the brief.

The game class instantiates 3 die objects and "BeginRound" rolls them and aggregates them. I really like the additional statistics features as a point of interest for the end user, and they could be useful in the event of expanding the game in the future.
You've added the ability to roll continuously for further rounds and I tested the input handling successfully.

You should add an error message in the event of an invalid input to explain the error for the user.

WonderboyMattT commented on CMP1903_A1_2324/Testing.cs:
This has been implemented properly.
Debug.Assert has been used to validate that the individual dice rolls are within the accepted range, that the sum of the dice rolls match the sum of the round, and that that sum is within the range of 3-18

WonderboyMattT commented on CMP1903_A1_2324/Program.cs:
You need to add test.RunTest() as currently the Testing class is instantiated but the method is never called. When testing, I altered the range of the die rolls to outside the accepted parameters and the game continued regardless.

WonderboyMattT reviewed (commented):
The provided code adheres to the C# style conventions, though there are minor exceptions.
Classes, methods, and variables all follow the correct naming conventions, though your values should be PascalCase, not camelCase.
There needs to be some robustness added to the error handling. For example, the "Roll()" method of the Die class, there is no explicit error handling for if the random number generator fails to generate a valid value. This is mitigated by the range given but it is good practice to prepare to handle unexpected occurrences.

NAbelSmith commented:
**Peer Review suggestions complete and implemented:**
- Implemented a new method RollDie() under Game Class, this performs the Roll() method for a given Die object.
- Implemented the way that dice are created & instantiated under Game Class, created a new property called DieList which is an Array of fixed size (defined by new private property _numOfDie - preset in code) and added a new constructor which adds the number of die set in _numOfDie to the array. This then caused a lot of other functions to be changed:
> - A new public method had to be created for the Testing Class to be able to fetch the private property _numOfDie so that it can also loop through the die and check rolls.
> - Places where the old method of each die being private Die _die1 and private Die _die2 etc were replaced with a for loop which then iterated through the array and performed the desired task on each die individually.
- Added more user friendly error messages to the Continue method under Game Class, this was done so that the user knows when their input is being rejected instead of simply repeating the input prompt.
- Fixed some unfound errors within Testing Class - when checking if the dice were in the desired range the logic was reversed, now addressed and fixed.

Copied from Pull Request #2
